aio_pika/robust_connection.py

```diff
@@ -95,11 +97,64 @@ async def __connection_attempt(
             closing = self.loop.create_future()
             closing.set_exception(e)
             await self.close_callbacks(closing)
-            await asyncio.gather(connection.close(e), return_exceptions=True)
+            await asyncio.gather(
+                transport.connection.close(e), return_exceptions=True,
+            )
             raise
-        return connection
+
+        await super()._on_connected(transport)
+
+        if self.connection_attempt:
+            await self.reconnect_callbacks()
+
+        self.connection_attempt += 1
+        self.__connection_close_event.clear()
+
+    async def __connection_factory(self) -> None:
+        log.debug("Starting connection factory for %r", self)
+        while not self.is_closed and not self._close_called:
+            log.debug("Waiting for connection close event for %r", self)
+            await self.__connection_close_event.wait()
+
+            if self.is_closed or self._close_called:
+                return
+
+            # noinspection PyBroadException
+            try:
+                self.transport = None
+                self.connected.clear()
+
+                log.debug("Connection attempt for %r", self)
+                await Connection.connect(self, self.__connect_timeout)
+
+                if not self.__fail_fast_future.done():
+                    self.__fail_fast_future.set_result(None)
+
+                log.debug("Connection made on %r", self)
+            except CONNECTION_EXCEPTIONS as e:
+                if not self.__fail_fast_future.done():
+                    self.__fail_fast_future.set_exception(e)
+                    return
+
+                log.warning(
+                    'Connection attempt to "%s" failed: %s. '
+                    "Reconnecting after %r seconds.",
+                    self, e, self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+            except Exception:
+                log.exception(
+                    "Reconnect attempt failed %s. "
+                    "Retrying after %r seconds.",
+                    self, self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+
+            await asyncio.sleep(self.reconnect_interval)
```

Comment:
I guess two sleeps above are redundant, are they?

Reply:
Nope previous one in the except block

Reply:
Yes, but this sleep will run after these `except` blocks.

Reply:
You right